### PR TITLE
Add a PE file section for determinism

### LINKED ISSUE #5940: Add a PE file section to indicate that the timestamp isn't a real timestamp
With the `/deterministic` flag, we place some bits into the PE file's timestamp field that isn't a timestamp. So that tools can tell that it isn't really a timestamp, we currently set the upper bit. We are also specifying an additional PE debug directory entry that, if present, is a specific indicator that the timestamp field contains a non-timestamp. We should produce this PE section when compiling to produce a deterministic output.

The following is the closest thing we have to a spec. I hope it makes sense:

> So here is the proposal on how we determine whether time stamp field in a PE actually store a real time date stamp.
 
> Like how we add incremental LTCG signature or COFF group info into PE, we can just define a new debug directory entry type.
 
> Here are all debug directory entry types that have already been defined:

> ```c 
#define IMAGE_DEBUG_TYPE_UNKNOWN          0
#define IMAGE_DEBUG_TYPE_COFF             1
#define IMAGE_DEBUG_TYPE_CODEVIEW         2
#define IMAGE_DEBUG_TYPE_FPO              3
#define IMAGE_DEBUG_TYPE_MISC             4
#define IMAGE_DEBUG_TYPE_EXCEPTION        5
#define IMAGE_DEBUG_TYPE_FIXUP            6
#define IMAGE_DEBUG_TYPE_OMAP_TO_SRC      7
#define IMAGE_DEBUG_TYPE_OMAP_FROM_SRC    8
#define IMAGE_DEBUG_TYPE_BORLAND          9
#define IMAGE_DEBUG_TYPE_RESERVED10       10
#define IMAGE_DEBUG_TYPE_CLSID            11
#define IMAGE_DEBUG_TYPE_VC_FEATURE       12
#define IMAGE_DEBUG_TYPE_POGO             13
#define IMAGE_DEBUG_TYPE_ILTCG            14
#define IMAGE_DEBUG_TYPE_MPX              15
```

> We can add a new one:

> ```c
#define IMAGE_DEBUG_TYPE_NO_TIMESTAMP  16
```

> Here is what debug directory looks like for ntdll.dll:

> ```none
        Time Type       Size      RVA  Pointer
    -------- ------ -------- -------- --------
    5609EE9F cv           22 0011DFE4   11AFE4    Format: RSDS, {FBA98142-7084-4ADE-9DDE-27DC9D6986ED}, 1, ntdll.pdb
    5609EE9F pogo        590 0011E01C   11B01C    50475500 (PGU)
```

> With repurposed time date stamp field in PE, the above will change to:

> ```none
        Time Type       Size      RVA  Pointer
    -------- ------ -------- -------- --------
    5609EE9F cv           22 0011DFE4   11AFE4    Format: RSDS, {FBA98142-7084-4ADE-9DDE-27DC9D6986ED}, 1, ntdll.pdb
    5609EE9F pogo        590 0011E01C   11B01C    50475500 (PGU)
    5609EE9F (16)        0  00000000   0 
```

> The last entry will indicate the time date stamp field in this binary doesn't represent a real time date stamp.

### PR DESCRIPTION
**Pull request by gafter**
to indicate that the timestamp isn't a real timestamp
Fixes #5940

@dotnet/roslyn-compiler Please review

### CODE REVIEW

File: Roslyn.sln [hunk 1]
```diff
@@ -134,7 +134,7 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "InteractiveWindow", "src\In
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "CSharpEditorServicesTest", "src\EditorFeatures\CSharpTest\CSharpEditorServicesTest.csproj", "{AC2BCEFB-9298-4621-AC48-1FF5E639E48D}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "CSharpEditorServicesTest2", "src\EditorFeatures\CSharpTest2\CSharpEditorServicesTest2.csproj", "{16e93074-4252-466c-89a3-3b905abaf779}"
```
Comment (AlekseyTs) [inline Roslyn.sln]: Are changes in this file intentional? Consider reverting otherwise.

File: src/Compilers/Core/Portable/PEWriter/PeWriter.cs [hunk 2]
```diff
@@ -1350,41 +1359,77 @@ private static void WriteSpaceForHash(Stream peStream, int strongNameSignatureSi
...
+            Debug.Assert(stamp.Length == 4);
+            writer.WriteBytes(stamp);
+            writer.WriteUInt32(version);
+            writer.WriteUInt32(debugType);
+            writer.WriteUInt32(sizeOfData);
+            writer.WriteUInt32(addressOfRawData);
+            writer.WriteUInt32(pointerToRawData);
+        }
+
+        private static byte[] zeroStamp = new byte[4]; // four bytes of zero
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/PEWriter/PeWriter.cs]: readonly?

File: src/Compilers/Core/Portable/PEWriter/PeWriter.cs [hunk 3]
```diff
@@ -1350,41 +1359,77 @@ private static void WriteSpaceForHash(Stream peStream, int strongNameSignatureSi
...
-            writer.WriteUInt32(0);
-
-            // PDB stamp & version
-            if (portablePdbContentId.IsDefault)
-            {
-                writer.WriteBytes(nativePdbContentId.Stamp);
-                writer.WriteUInt32(0);
-            }
-            else
+            const int ImageDebugTypeCodeView = 2;
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/PEWriter/PeWriter.cs]: Invalid capitalization for a local's name.

File: src/Compilers/Core/Portable/PEWriter/PeWriter.cs [hunk 4]
```diff
@@ -1350,41 +1359,77 @@ private static void WriteSpaceForHash(Stream peStream, int strongNameSignatureSi
...
+                debugType: ImageDebugTypeCodeView,
+                sizeOfData: (uint)dataSize,
+                addressOfRawData: (uint)textSection.RelativeVirtualAddress + dataOffset, // RVA of the data
+                pointerToRawData: (uint)textSection.PointerToRawData + dataOffset); // position of the data in the PE stream
+
+            if (this._deterministic)
             {
-                writer.WriteBytes(portablePdbContentId.Stamp);
-                writer.WriteUInt32('P' << 24 | 'M' << 16 | 0x01 << 8 | 0x00);
+                const int ImageDebugTypeDeterministicTimestamp = 16;
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/PEWriter/PeWriter.cs]: Invalid capitalization.

**Comment by AlekseyTs** [comment]
CC @tmat

**Comment by AlekseyTs** [comment]
LGTM

File: Roslyn.sln (hunk 1 above)
Comment (gafter) [inline Roslyn.sln]: I keep reverting and it keeps changing. It appears that VS wants to make the GUIDs consistently use uppercase.

File: src/Compilers/Core/Portable/PEWriter/PeWriter.cs (hunk 3 above)
Comment (gafter) [inline src/Compilers/Core/Portable/PEWriter/PeWriter.cs]: That is how it is spelled in the specification.

File: src/Compilers/Core/Portable/PEWriter/PeWriter.cs (hunk 4 above)
Comment (gafter) [inline src/Compilers/Core/Portable/PEWriter/PeWriter.cs]: Taken from the specification.

File: src/Compilers/Core/Portable/PEWriter/PeWriter.cs [hunk 5]
```diff
@@ -461,14 +461,23 @@ private int ComputeOffsetToMetadata(int ilStreamLength)
...
+                    sizeof(uint) +   // Type
+                    sizeof(uint) +   // SizeOfData
+                    sizeof(uint) +   // AddressOfRawData
+                    sizeof(uint);    // PointerToRawData
+
+        /// <summary>
+        /// The size of our debug directory: one entry for debug information, and an optional second one indicating
+        /// that the timestamp is deterministic (i.e. not really a timestamp)
+        /// </summary>
+        private int ImageDebugDirectoryBaseSize => (_deterministic ? 2 : 1) * ImageDebugDirectoryEntrySize;
```
Comment (gafter) [inline src/Compilers/Core/Portable/PEWriter/PeWriter.cs]: This isn't right. We now need to produce the debug table when debug is enabled (which adds one entry plus some data) or when determinism is enabled (which adds one entry). Fixing this will require changes in a few places.

File: Roslyn.sln (hunk 1 above)
Comment (AlekseyTs) [inline Roslyn.sln]: Still, consider excluding this file from the commit because this change is not related to the work you are doing. If you want, submit a separate PR for this file.

File: src/Compilers/Core/Portable/PEWriter/PeWriter.cs (hunk 3 above)
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/PEWriter/PeWriter.cs]: Sorry for not being clear. Names of locals shouldn't start with capital letters.

File: src/Compilers/Core/Portable/PEWriter/PeWriter.cs (hunk 4 above)
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/PEWriter/PeWriter.cs]: Names of locals shouldn't start with capital letters.

File: src/Compilers/Core/Portable/PEWriter/PeWriter.cs [hunk 6]
```diff
@@ -1350,41 +1359,77 @@ private static void WriteSpaceForHash(Stream peStream, int strongNameSignatureSi
...
+            writer.WriteUInt32(addressOfRawData);
+            writer.WriteUInt32(pointerToRawData);
+        }
+
+        private static byte[] zeroStamp = new byte[4]; // four bytes of zero
+
+        /// <summary>
+        /// Write the entire "Debug Directory (Image Only)" along with data that it points to.
+        /// </summary>
+        /// <param name="peStream"></param>
```
Comment (tmat) [inline src/Compilers/Core/Portable/PEWriter/PeWriter.cs]: Remove <param> comments since they are empty.

**Comment by tmat** [comment]
:+1:

**Comment by gafter** [comment]
@dotnet-bot Test mac/dbg please

**Comment by gafter** [comment]
@dotnet-bot test this please
